File: cpp/src/153.cpp
#include <vector>

using std::vector;

class Solution {
  int findMin(vector<int> &nums) {

    size_t low = 0;
    size_t high = nums.size() - 1;

    while (low < high) {
      size_t mid = low + (high - low) / 2;
      if (nums[mid] > nums[high]) {
        low = mid + 1;
      } else {
        high = mid;
      }
    }
    return nums[low];
  }
};

File: python/src/153.py
from typing import List


def findMin(nums: List[int]) -> int:
    low = 0
    high = len(nums) - 1

    while low < high:
        mid = low + (high - low) // 2
        print(low, mid, high)

        if nums[mid] < nums[high]:
            low = mid + 1
        else:
            high = mid

    print(low)
    return nums[low]


if __name__ == "__main__":
    assert findMin([1, 2, 3, 4, 5]) == 5
    assert findMin([4, 5, 1, 2, 3]) == 1
